plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/EqualityDeleteFilter.java

```diff
@@ -17,52 +17,182 @@
+        }
+
+        public void setLoaded()
+        {
+            synchronized (lockToken) {
+                this.isLoaded = true;
+            }
+        }
+
+        public void releaseLoading()
+        {
+            synchronized (lockToken) {
+                isLoading = false;
+            }
+        }
+    }
+
+    private final Schema deleteSchema;
+    private final StructLikeMap<DataSequenceNumber> deleteMap;
+    private final HashMap<String, FileToken> loadedFiles;
+    private final Object containerReference;
+
+    /**
+     * @param containerReference This is passed in since we don't want the weak-reference holding this objects container to be nulled out while this
+     * equality delete filter is still in use.
+     */
+    public EqualityDeleteFilter(Schema deleteSchema, Object containerReference)
+    {
+        this.deleteSchema = requireNonNull(deleteSchema, "deleteSchema is null");
+        this.deleteMap = StructLikeMap.create(deleteSchema.asStruct());
+        this.loadedFiles = new HashMap<>();
+        this.containerReference = containerReference;
+    }
+
+    public FileToken getLoadFileToken(String fileName, long deleteFileSequenceNumber)
+    {
+        synchronized (loadedFiles) {
+            FileToken fileToken = loadedFiles.get(fileName);
+            if (fileToken == null) {
+                fileToken = new FileToken(deleteFileSequenceNumber);
+                loadedFiles.put(fileName, fileToken);
+            }
+            return fileToken;
+        }
     }
 
     @Override
-    public RowPredicate createPredicate(List<IcebergColumnHandle> columns)
+    public RowPredicate createPredicate(List<IcebergColumnHandle> columns, long splitDataSequenceNumber)
     {
         Type[] types = columns.stream()
                 .map(IcebergColumnHandle::getType)
                 .toArray(Type[]::new);
 
-        return (page, position) -> {
-            StructLike row = new LazyTrinoRow(types, page, position);
-            return !deletedRows.test(row);
-        };
+        Schema fileSchema = schemaFromHandles(columns);
+        boolean hasRequiredColumns = columns.size() >= fileSchema.columns().size();
+        for (Types.NestedField column : deleteSchema.columns()) {
+            hasRequiredColumns = hasRequiredColumns && fileSchema.findField(column.fieldId()) != null;
+        }
+        if (!hasRequiredColumns) {
+            // If we don't have all the required columns this delete filter can't be applied.
+            // The iceberg split manager is responsible for making sure that we have all the required columns when a delete filter should be applied
+            return (page, position) -> true;
+        }
+        else {
+            StructProjection projection = StructProjection.create(fileSchema, deleteSchema);
+
+            return (page, position) -> {
+                StructProjection wrapped = projection.wrap(new LazyTrinoRow(types, page, position));
+                synchronized (deleteMap) {
```

Comment:
we probably need a read-safe thread map or a read-safe clone here. Otherwise contention would be a big issue.